tsbrowse/preprocess.py

```diff
@@ -0,0 +1,546 @@
+    """
+    tree_pos = alloc_tree_position(ts)
+    return _compute_per_tree_stats(
+        tree_pos,
+        ts.num_trees,
+        ts.num_nodes,
+        ts.nodes_time,
+        ts.edges_parent,
+        ts.edges_child,
+    )
+
+
+def trees(ts):
+    num_trees = ts.num_trees
+    total_branch_length, num_internal_nodes, max_arity = compute_per_tree_stats(ts)
+
+    # FIXME - need to add this to the computation above
+    mean_internal_arity = np.zeros(num_trees)
+
+    site_tree_index = (
+        np.searchsorted(ts.breakpoints(as_array=True), ts.sites_position, side="right")
+        - 1
+    )
+    unique_values, counts = np.unique(site_tree_index, return_counts=True)
+    sites_per_tree = np.zeros(ts.num_trees, dtype=np.int64)
+    sites_per_tree[unique_values] = counts
+
+    mutation_tree_index = site_tree_index[ts.mutations_site]
+    unique_mutation_values, mutation_counts = np.unique(
+        mutation_tree_index, return_counts=True
+    )
+    mutations_per_tree = np.zeros(ts.num_trees, dtype=np.int64)
+    mutations_per_tree[unique_mutation_values] = mutation_counts
+
+    breakpoints = ts.breakpoints(as_array=True)
+    logger.info("Pre processed trees")
+    return {
+        "left": breakpoints[:-1],
+        "right": breakpoints[1:],
+        "total_branch_length": total_branch_length,
+        "mean_internal_arity": mean_internal_arity,
+        "max_internal_arity": max_arity,
+        "num_sites": sites_per_tree,
+        "num_mutations": mutations_per_tree,
+    }
+
+
+def preprocess(tszip_path, output_path=None):
+    tszip_path = pathlib.Path(tszip_path)
+    if output_path is None:
+        output_path = tszip_path.with_suffix(".tsbrowse")
+    preprocessors = [mutations, nodes, trees, edges, sites]
+    with tqdm(total=2 + len(preprocessors), desc="Processing") as pbar:
+        ts = tszip.load(tszip_path)
+        pbar.update(1)
+
+        tszip.compress(ts, output_path)
+        pbar.update(1)
+
+        # Preprocess the data first so we can error out before writing to the file
+        data = {}
+        with ProcessPoolExecutor() as executor:
+            futures = {
+                executor.submit(preprocessor, ts): preprocessor.__name__.split(".")[-1]
+                for preprocessor in preprocessors
+            }
+            for future in as_completed(futures):
+                group_name = futures[future]
+                data[group_name] = future.result()
+                pbar.update(1)
+
+    with zarr.ZipStore(output_path, mode="a") as zarr_store:
+        root = zarr.group(store=zarr_store)
+        total_arrays = sum(len(arrays) for arrays in data.values())
+        with tqdm(total=total_arrays, desc="Writing") as pbar:
+            for table_name, arrays in data.items():
+                for array_name, array in arrays.items():
+                    root[f"{table_name}/{array_name}"] = array
+                    pbar.update(1)
+        root.attrs["tsbrowse_data_version"] = TSBROWSE_DATA_VERSION
```

Comment:
Let's make this ``root.attrs["tsbrowse"] = {"data_version": TSBROWSE_DATA_VERSION}`` so that we can add other stuff in there later if we want.